File: Bwl.Framework/src/AutoUI/AutoUI/AutoUIDisplay.vb
Imports Bwl.Framework

Public Class AutoUIDisplay
    Public Event AutoFormDescriptorUpdated(sender As AutoUIDisplay)
    Public Property AutoFormDescriptor As RemoteAutoFormDescriptor

    Private WithEvents _ui As IAutoUI
    Private _loaded As Boolean

    Public Sub New()
        InitializeComponent()
    End Sub

    Public Property ConnectedUI As IAutoUI
        Get
            Return _ui
        End Get
        Set(value As IAutoUI)
            _ui = value
            If _ui IsNot Nothing Then
                If Me.IsDisposed Then Return
                If _loaded Then _ui.GetBaseInfos()
            End If
        End Set
    End Property

    Private Sub AutoUIDisplay_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If Me.IsDisposed Then Return
        _loaded = True
    End Sub

    Private Sub RemoveControls()
        If Me.IsDisposed Then Return
        Me.Invoke(Sub() panel.Controls.Clear())
    End Sub

    Public Sub RecreateControls()
        If Me.IsDisposed Then Return
        _ui.GetBaseInfos()
    End Sub

    Private Sub _ui_RequestToSend(id As String, dataname As String, data() As Byte) Handles _ui.RequestToSend
        If Me.IsDisposed Then Return
        Try
            If AutoFormDescriptor IsNot Nothing AndAlso id = AutoFormDescriptor.Info.ID Then
                AutoFormDescriptor.ProcessData(dataname, data)
            End If

            For i = 0 To panel.Controls.Count - 1
                Dim ctl = DirectCast(panel.Controls.Item(i), BaseRemoteElement)
                If ctl.Info.ID.ToLower = id.ToLower Then
                    If dataname = "base-info-change" Then
                        ctl.Info.SetFromBytes(data)
                    Else
                        ctl.ProcessData(dataname, data)
                    End If
                End If
            Next
        Catch ex As Exception
        End Try
    End Sub


    Private Sub _ui_BaseInfosReady(infos As Byte()()) Handles _ui.BaseInfosReady
        If Me.IsDisposed Then Return
        Try
            RemoveControls()
            For Each infoBytes In infos
                Dim info = UIElementInfo.CreateFromBytes(infoBytes)
                Dim ctl As BaseRemoteElement = Nothing
                Select Case info.Type
                    Case "AutoButton"
                        ctl = New RemoteAutoButton(info)
                    Case "AutoBitmap"
                        ctl = New RemoteAutoBitmap(info)
                    Case "AutoTextbox"
                        ctl = New RemoteAutoTextbox(info)
                    Case "AutoListbox"
                        ctl = New RemoteAutoListbox(info)
                    Case "AutoListbox"
                        ctl = New RemoteAutoListbox(info)
                    Case "AutoFormDescriptor"
                        _AutoFormDescriptor = New RemoteAutoFormDescriptor(info)
                        AddHandler _AutoFormDescriptor.Updated, Sub()
                                                                    RaiseEvent AutoFormDescriptorUpdated(Me)
                                                                End Sub
                        AddHandler _AutoFormDescriptor.RequestToSend, Sub(source As IUIElement, dataname As String, data As Byte())
                                                                          _ui.ProcessData(source.Info.ID, dataname, data)
                                                                      End Sub
                        _AutoFormDescriptor.Update()
                End Select
                If ctl IsNot Nothing Then
                    AddHandler ctl.RequestToSend, Sub(source As IUIElement, dataname As String, data As Byte())
                                                      _ui.ProcessData(source.Info.ID, dataname, data)
                                                  End Sub
                    Me.Invoke(Sub() panel.Controls.Add(ctl))
                End If
            Next
        Catch ex As Exception
        End Try
    End Sub
End Class

File: Framework/Framework/Bwl.Framework.Windows/AutoUI/AutoUI/AutoUIDisplay.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.Versioning;
using System.Windows.Forms;

namespace Bwl.Framework.Windows
{

    [SupportedOSPlatform("windows")]
    public partial class AutoUIDisplay
    {
        public event EventHandler<AutoUIDisplay> AutoFormDescriptorUpdated;

        public RemoteAutoFormDescriptor AutoFormDescriptor { get; set; }

        private IAutoUI __ui;

        private IAutoUI _ui
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return __ui;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (__ui != null)
                {
                    __ui.RequestToSend -= _ui_RequestToSend;
                    __ui.BaseInfosReady -= _ui_BaseInfosReady;
                }

                __ui = value;
                if (__ui != null)
                {
                    __ui.RequestToSend += _ui_RequestToSend;
                    __ui.BaseInfosReady += _ui_BaseInfosReady;
                }
            }
        }
        private bool _loaded;

        public AutoUIDisplay()
        {
            InitializeComponent();
            Application.EnableVisualStyles();
        }

        public IAutoUI ConnectedUI
        {
            get
            {
                return _ui;
            }
            set
            {
                _ui = value;
                if (_ui is not null)
                {
                    if (IsDisposed)
                        return;
                    if (_loaded)
                        _ui.GetBaseInfos();
                }
            }
        }

        private void AutoUIDisplay_Load(object sender, EventArgs e)
        {
            if (IsDisposed)
                return;
            _loaded = true;
        }

        private void RemoveControls()
        {
            if (IsDisposed)
                return;
            Invoke(() => panel.Controls.Clear());
        }

        public void RecreateControls()
        {
            if (IsDisposed)
                return;
            _ui.GetBaseInfos();
        }

        private void _ui_RequestToSend(string id, string dataname, byte[] data)
        {
            if (IsDisposed)
                return;
            try
            {
                if (AutoFormDescriptor is not null && (id ?? "") == (AutoFormDescriptor.Info.ID ?? ""))
                {
                    AutoFormDescriptor.ProcessData(dataname, data);
                }

                for (int i = 0, loopTo = panel.Controls.Count - 1; i <= loopTo; i++)
                {
                    BaseRemoteElement ctl = (BaseRemoteElement)panel.Controls[i];
                    if ((ctl.Info.ID.ToLower() ?? "") == (id.ToLower() ?? ""))
                    {
                        if (dataname == "base-info-change")
                        {
                            ctl.Info.SetFromBytes(data);
                        }
                        else
                        {
                            ctl.ProcessData(dataname, data);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
            }
        }


        private void _ui_BaseInfosReady(byte[][] infos)
        {
            if (IsDisposed)
                return;
            try
            {
                RemoveControls();
                foreach (var infoBytes in infos)
                {
                    var info = UIElementInfo.CreateFromBytes(infoBytes);
                    BaseRemoteElement? ctl = null; // Fix for CS8600: Use nullable type

                    switch (info.Type ?? string.Empty) // Use string.Empty for null coalescing
                    {
                        case nameof(AutoImage):
                            ctl = new RemoteAutoImage(info);
                            break;

                        case nameof(AutoButton):
                            ctl = new RemoteAutoButton(info);
                            break;

                        case nameof(AutoTextbox):
                            ctl = new RemoteAutoTextbox(info);
                            break;

                        case nameof(AutoListbox):
                            ctl = new RemoteAutoListbox(info);
                            break;

                        case nameof(AutoFormDescriptor):
                            AutoFormDescriptor = new RemoteAutoFormDescriptor(info);
                            AutoFormDescriptor.Updated += (object? sender, RemoteAutoFormDescriptor descriptor) => AutoFormDescriptorUpdated?.Invoke(this, this);
                            AutoFormDescriptor.RequestToSend += (sender, e) => _ui.ProcessData(e.source.Info.ID, e.dataname, e.data);
                            AutoFormDescriptor.Update();
                            break;
                    }

                    if (ctl is not null)
                    {
                        ctl.RequestToSend += (sender, e) => _ui.ProcessData(e.source.Info.ID, e.dataname, e.data);
                        Invoke(() => panel.Controls.Add(ctl));
                    }
                }
            }
            catch (Exception ex)
            {
                // Log the exception or handle it appropriately
                Console.WriteLine($"An error occurred: {ex.Message}");
            }
        }
    }
}